### PR TITLE
conn: new package that splits out the Bind and Endpoint types

### PR DESCRIPTION
**Pull request by crawshaw**
The sticky socket code stays in the device package for now,
as it reaches deeply into the peer list.

This is the first step in an effort to split some code out of
the very busy device package.

Signed-off-by: David Crawshaw <crawshaw@tailscale.com>

### CODE REVIEW

File: conn/conn.go [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+	"strings"
+)
+
+const (
+	ConnRoutineNumber = 2
+)
+
+// A Bind listens on a port for both IPv6 and IPv4 UDP traffic.
+type Bind interface {
+	LastMark() uint32
```
Comment (bradfitz) [inline conn/conn.go]: docs on all these would be nice (including concurrency properties) if you know the answers.

File: conn/conn_linux.go [hunk 2]
```diff
@@ -111,47 +103,16 @@ func CreateEndpoint(s string) (Endpoint, error) {
...
-	if err != nil {
-		unix.Close(sock)
-		return -1, err
-	}
-	return sock, nil
-
-}
-
-func CreateBind(port uint16, device *Device) (*nativeBind, uint16, error) {
+func CreateBind(port uint16, device interface{}) (*nativeBind, uint16, error) {
```
Comment (bradfitz) [inline conn/conn_linux.go]: exported func returning unexported concrete type is weird.

Comment (bradfitz) [reply conn/conn_linux.go]: and what's the returned uint16?

document?

**Comment by crawshaw** [comment]
I figure each of these commits going upstream should leave things better than it was. I've tried to add some docs, but clearly there's more that could be done.

**Comment by crawshaw** [comment]
(Anyway PTAL.)

**Comment by crawshaw** [comment]
Moved to https://github.com/WireGuard/wireguard-go/pull/9.

**Pull request closed without merging**